backend/src/nodes/ncnn_nodes.py

```diff
@@ -19,103 +16,49 @@
-        self.output_name = output_name
-
-
 @NodeFactory.register("chainner:ncnn:load_model")
 class NcnnLoadModelNode(NodeBase):
     def __init__(self):
         super().__init__()
         self.description = "Load NCNN model (.bin and .param files)."
         self.inputs = [ParamFileInput(), BinFileInput()]
-        self.outputs = [NcnnNetOutput(), TextOutput("Model Name")]
+        self.outputs = [NcnnModelOutput(), TextOutput("Model Name")]
 
         self.category = NCNNCategory
         self.name = "Load Model"
         self.icon = "NCNN"
         self.sub = "Input & Output"
 
-    def get_param_info(self, param_path):
-        input_name = "data"
-        output_name = "output"
-        out_nc = 3
-
-        with open(param_path, encoding="utf-8") as f:
-            lines = f.read()
-
-            assert (
-                not "MemoryData" in lines
-            ), "This NCNN param file contains invalid layers"
-
-            # # Scale
-            # regex = 'Interp\s*\w*\s*.*2=(\d.?\d*)'
-            # matches = re.findall(regex, lines)
-            # scale = int(np.prod([float(n) for n in matches]))
-
-            # Input name
-            regex = r"Input\s+([\w.]+)\s+0\s1\s(\w+)"
-            matches = re.findall(regex, lines)
-            if len(matches) > 0:
-                if any(isinstance(el, tuple) for el in matches):
-                    matches = matches[-1]
-                _, input_name = matches
-
-            # Output name & out nc
-            # regex = '\w+\s+([\w.]+)\s+\d+\s+\d+\s+\d+\s+([^\d\s]+)\s0=(\d)'
-            regex = r"\s([^\s]+)\s0=(\d)"
-            matches = re.findall(regex, lines)
-            if len(matches) > 0:
-                if any(isinstance(el, tuple) for el in matches):
-                    matches = matches[-1]
-                output_name, out_nc = matches
-        logger.info(f"{input_name}, {output_name}, {out_nc}")
-
-        return input_name, output_name, out_nc
-
-    def run(self, param_path: str, bin_path: bytes) -> Tuple[NcnnNetData, str]:
+    def run(self, param_path: str, bin_path: str) -> Tuple[NcnnModel, str]:
         assert os.path.exists(
             param_path
         ), f"Param file at location {param_path} does not exist"
         assert os.path.exists(
             bin_path
-        ), f"Bin file at location {param_path} does not exist"
+        ), f"Bin file at location {bin_path} does not exist"
 
         assert os.path.isfile(param_path), f"Path {param_path} is not a file"
-        assert os.path.isfile(bin_path), f"Path {param_path} is not a file"
-
-        input_name, output_name, _out_nc = self.get_param_info(param_path)
+        assert os.path.isfile(bin_path), f"Path {bin_path} is not a file"
 
+        model = NcnnModel()
+        model.load_model(param_path, bin_path)
         with open(bin_path, "rb") as f:
-            bin_file_data = f.read()
-        bin_data = parse_ncnn_bin_from_buffer(bin_file_data)
-
+            model.weights_bin = f.read()
```

Comment:
The blank constructor exists because when converting ONNX to NCNN, the layers are being converted and added to it one at a time. I could overload the constructor to pass files to it directly, though.

As for load_model, it's loading the param layer information, as well as the corresponding numerical data arrays from the bin. For our purposes, we need the weights as arrays so we can interpolate them, but the actual ncnn load_model needs raw byte data. The whole having a separate attribute for the bin in bytes is a speed hack on my part, because converting the numerical arrays back to bytes takes time (especially with python for loops being so slow). When loading the model from file, the fastest way to do that is just read it straight from the bin. When converting from onnx or interpolating, it converts the arrays to bytes as it goes, then concatenates the byte strings at the end.

Reply:
Apparently python does not do overloading lol